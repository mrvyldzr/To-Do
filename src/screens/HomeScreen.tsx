// React Imports
import React, {FC, useContext, useEffect, useState} from 'react'

// React Native Imports
import {View, Text, StyleSheet, Image, TouchableOpacity} from 'react-native'

// Constant Imports
import {COLOURS} from '../constants'
import Images from '../constants/Images'

// Component Imports
import {List} from '../components'

// Database Imports
import database from '../database/database'

// Model Imports
import {Task} from './models'

// Translation Imports
import {TranslationContext} from '../translator/providers'

interface HomeScreenProps {
  navigation?: any
}

export const HomeScreen: FC<HomeScreenProps> = ({navigation}) => {
  const [tasks, setTasks] = useState<Task[]>([])
  const {translate} = useContext(TranslationContext)

  useEffect(() => {
    loadTasks()
  }, [tasks])

  const loadTasks = () => {
    database
      .getAllTasks()
      .then((data: Task[]) => setTasks(data))
      .catch((error: any) => console.error(translate('DATABASE_ERROR_FETCHING_DATA'), error))
  }

  const handleItemPress = (id: number) => {
    navigation.navigate('Detail', {taskId: id})
  }

  const handleToggleComplete = (id: number, completed: boolean) => {
    database
      .updateTask(id, {completed})
      .then(() => {
        loadTasks()
      })
      .catch((error) => {
        console.error('DATABASE_ERROR_UPDATING_TASK', error)
      })
  }

  const renderItem = ({item}: {item: Task}) => (
    <TouchableOpacity onPress={() => handleItemPress(item.id)}>
      <View
        style={[
          styles.itemContainer,
          {
            backgroundColor: item.completed ? COLOURS.RED_LIGHT : COLOURS.ORANGE_LIGHT,
          },
        ]}>
        <View style={styles.titleItem}>
          <Text style={styles.titleText} numberOfLines={1}>
            {item.title}
          </Text>
          <TouchableOpacity onPress={() => handleToggleComplete(item.id, !item.completed)}>
            <Image
              source={item.completed ? Images.CHECKED : Images.UNCHECKED}
              style={styles.checkIcon}
            />
          </TouchableOpacity>
        </View>

        <Text style={styles.contentText} numberOfLines={2}>
          {item.content}
        </Text>

        <View style={{paddingTop: 4}}>
          <Text style={styles.contentText}>
            {item.date && (
              <Text>
                {new Date(item.date).getDate()}.{new Date(item.date).getMonth() + 1}.
                {new Date(item.date).getFullYear()} {new Date(item.date).toLocaleTimeString()}
              </Text>
            )}
          </Text>
        </View>
      </View>
    </TouchableOpacity>
  )

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Image source={Images.TO_DO_TITLE_IMAGE} />
        <Image source={Images.SETTINGS} />
      </View>

      <View style={styles.subHeader}>
        <View style={styles.subHeaderItem}>
          <Image source={Images.UNION} />
          <Image source={Images.TO_DO_LIST_IMAGE} />
        </View>
        <Image source={Images.FILTER} />
      </View>

      <View style={styles.flatListContainer}>
        <List
          data={tasks}
          renderItem={renderItem}
          keyExtractor={(item: Task) => item.id.toString()}
        />
      </View>

      <View style={styles.addButton}>
        <TouchableOpacity onPress={() => navigation.navigate('AddTask', {taskId: null})}>
          <Image source={Images.PLUS_CIRCLE} />
        </TouchableOpacity>
      </View>
    </View>
  )
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
  },
  header: {
    paddingHorizontal: 24,
    paddingTop: 120,
    paddingBottom: 48,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  subHeader: {
    paddingHorizontal: 24,
    paddingBottom: 48,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  subHeaderItem: {
    flexDirection: 'row',
    gap: 16,
  },
  flatListContainer: {
    flex: 1,
    paddingHorizontal: 24,
  },
  itemContainer: {
    width: 340,
    height: 130,
    padding: 20,
    borderRadius: 8,
    marginBottom: 10,
  },
  titleItem: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  titleText: {
    fontSize: 24,
    fontWeight: 'bold',
    color: COLOURS.WHITE,
  },
  contentText: {
    fontSize: 14,
    marginTop: 8,
    color: COLOURS.WHITE,
  },
  addButton: {
    flex: 0.1,
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    marginBottom: 80,
    marginRight: 20,
  },
  checkIcon: {
    width: 20,
    height: 20,
    marginLeft: 5,
  },
})
