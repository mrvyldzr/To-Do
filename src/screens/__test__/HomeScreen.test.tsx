// React Imports
import React from 'react'

// React Native Imports
import {render, fireEvent} from '@testing-library/react-native'

// Component Imports
import {HomeScreen} from '../HomeScreen'

describe('HomeScreen', () => {
  it('renders HomeScreen correctly', () => {
    const renderedHomeScreen = render(<HomeScreen />)

    expect(renderedHomeScreen).toMatchSnapshot()
  })
})
