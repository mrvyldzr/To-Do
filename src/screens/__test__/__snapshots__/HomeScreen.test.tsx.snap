// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders HomeScreen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 48,
        "paddingHorizontal": 24,
        "paddingTop": 120,
      }
    }
  >
    <Image
      source={1}
    />
    <Image
      source={1}
    />
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 48,
        "paddingHorizontal": 24,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 16,
        }
      }
    >
      <Image
        source={1}
      />
      <Image
        source={1}
      />
    </View>
    <Image
      source={1}
    />
  </View>
  <View
    style={
      {
        "flex": 1,
        "paddingHorizontal": 24,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RCTScrollView
        applyWindowCorrection={[Function]}
        canChangeSize={true}
        contentContainerStyle={
          {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 0,
            "paddingRight": 0,
          }
        }
        contentHeight={0}
        contentWidth={0}
        data={[]}
        dataProvider={
          DataProvider {
            "_data": [],
            "_firstIndexToProcess": 0,
            "_hasStableIds": true,
            "_requiresDataChangeHandling": false,
            "_size": 0,
            "getStableId": [Function],
            "rowHasChanged": [Function],
          }
        }
        disableRecycling={false}
        estimatedItemSize={100}
        extendedState={{}}
        externalScrollView={[Function]}
        finalRenderAheadOffset={250}
        forceNonDeterministicRendering={true}
        horizontal={false}
        initialOffset={0}
        initialRenderIndex={0}
        isHorizontal={false}
        keyExtractor={[Function]}
        layoutProvider={
          GridLayoutProviderWithProps {
            "_acceptableRelayoutDelta": 1,
            "_getHeightOrWidth": [Function],
            "_getLayoutTypeForIndex": [Function],
            "_getSpan": [Function],
            "_hasExpired": false,
            "_maxSpan": 1,
            "_setLayoutForType": [Function],
            "_tempDim": {
              "height": 0,
              "width": 0,
            },
            "averageWindow": AverageWindow {
              "currentAverage": 100,
              "currentCount": 1,
              "inputValues": [
                100,
              ],
              "nextIndex": 1,
            },
            "defaultEstimatedItemSize": 100,
            "layoutObject": {
              "size": undefined,
              "span": undefined,
            },
            "props": {
              "data": [],
              "estimatedItemSize": 100,
              "keyExtractor": [Function],
              "numColumns": 1,
              "renderItem": [Function],
            },
            "renderWindowInsets": {
              "height": 0,
              "width": 0,
            },
            "shouldRefreshWithAnchoring": true,
          }
        }
        maxRenderAhead={750}
        numColumns={1}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onEndReachedThresholdRelative={0}
        onItemLayout={[Function]}
        onLayout={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onSizeChanged={[Function]}
        onVisibleIndicesChanged={[Function]}
        removeClippedSubviews={false}
        renderAheadOffset={0}
        renderAheadStep={250}
        renderContentContainer={[Function]}
        renderItem={[Function]}
        renderItemContainer={[Function]}
        rowRenderer={[Function]}
        scrollEventThrottle={16}
        scrollThrottle={16}
        scrollViewProps={
          {
            "contentContainerStyle": {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
              "paddingLeft": 0,
              "paddingRight": 0,
            },
            "onLayout": [Function],
            "onScrollBeginDrag": [Function],
            "refreshControl": undefined,
            "style": {
              "minHeight": 1,
              "minWidth": 1,
            },
          }
        }
        style={
          {
            "minHeight": 1,
            "minWidth": 1,
          }
        }
        suppressBoundedSizeException={true}
        windowCorrectionConfig={
          {
            "applyToInitialOffset": true,
            "applyToItemScroll": true,
            "value": {
              "endCorrection": 0,
              "startCorrection": 0,
              "windowShift": -0,
            },
          }
        }
      >
        <View>
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <View
              style={
                {
                  "paddingLeft": 0,
                  "paddingTop": 0,
                }
              }
            />
            <View
              style={
                [
                  undefined,
                  undefined,
                ]
              }
            />
            <AutoLayoutView
              enableInstrumentation={false}
              horizontal={false}
              onBlankAreaEvent={[Function]}
              onLayout={[Function]}
              renderAheadOffset={0}
              scrollOffset={0}
              style={
                {
                  "height": 0,
                  "width": 0,
                }
              }
              windowSize={0}
            />
            <CellContainer
              index={-1}
              style={
                [
                  undefined,
                  undefined,
                ]
              }
            />
            <View
              style={
                {
                  "paddingBottom": 0,
                  "paddingRight": 0,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flex": 0.1,
        "justifyContent": "flex-end",
        "marginBottom": 80,
        "marginRight": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Image
        source={1}
      />
    </View>
  </View>
</View>
`;

exports[`renders HomeScreen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 48,
        "paddingHorizontal": 24,
        "paddingTop": 120,
      }
    }
  >
    <Image
      source={1}
    />
    <Image
      source={1}
    />
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 48,
        "paddingHorizontal": 24,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 16,
        }
      }
    >
      <Image
        source={1}
      />
      <Image
        source={1}
      />
    </View>
    <Image
      source={1}
    />
  </View>
  <View
    style={
      {
        "flex": 1,
        "paddingHorizontal": 24,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RCTScrollView
        applyWindowCorrection={[Function]}
        canChangeSize={true}
        contentContainerStyle={
          {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 0,
            "paddingRight": 0,
          }
        }
        contentHeight={0}
        contentWidth={0}
        data={[]}
        dataProvider={
          DataProvider {
            "_data": [],
            "_firstIndexToProcess": 0,
            "_hasStableIds": true,
            "_requiresDataChangeHandling": false,
            "_size": 0,
            "getStableId": [Function],
            "rowHasChanged": [Function],
          }
        }
        disableRecycling={false}
        estimatedItemSize={100}
        extendedState={{}}
        externalScrollView={[Function]}
        finalRenderAheadOffset={250}
        forceNonDeterministicRendering={true}
        horizontal={false}
        initialOffset={0}
        initialRenderIndex={0}
        isHorizontal={false}
        keyExtractor={[Function]}
        layoutProvider={
          GridLayoutProviderWithProps {
            "_acceptableRelayoutDelta": 1,
            "_getHeightOrWidth": [Function],
            "_getLayoutTypeForIndex": [Function],
            "_getSpan": [Function],
            "_hasExpired": false,
            "_maxSpan": 1,
            "_setLayoutForType": [Function],
            "_tempDim": {
              "height": 0,
              "width": 0,
            },
            "averageWindow": AverageWindow {
              "currentAverage": 100,
              "currentCount": 1,
              "inputValues": [
                100,
              ],
              "nextIndex": 1,
            },
            "defaultEstimatedItemSize": 100,
            "layoutObject": {
              "size": undefined,
              "span": undefined,
            },
            "props": {
              "data": [],
              "estimatedItemSize": 100,
              "keyExtractor": [Function],
              "numColumns": 1,
              "renderItem": [Function],
            },
            "renderWindowInsets": {
              "height": 0,
              "width": 0,
            },
            "shouldRefreshWithAnchoring": true,
          }
        }
        maxRenderAhead={750}
        numColumns={1}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onEndReachedThresholdRelative={0}
        onItemLayout={[Function]}
        onLayout={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onSizeChanged={[Function]}
        onVisibleIndicesChanged={[Function]}
        removeClippedSubviews={false}
        renderAheadOffset={0}
        renderAheadStep={250}
        renderContentContainer={[Function]}
        renderItem={[Function]}
        renderItemContainer={[Function]}
        rowRenderer={[Function]}
        scrollEventThrottle={16}
        scrollThrottle={16}
        scrollViewProps={
          {
            "contentContainerStyle": {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
              "paddingLeft": 0,
              "paddingRight": 0,
            },
            "onLayout": [Function],
            "onScrollBeginDrag": [Function],
            "refreshControl": undefined,
            "style": {
              "minHeight": 1,
              "minWidth": 1,
            },
          }
        }
        style={
          {
            "minHeight": 1,
            "minWidth": 1,
          }
        }
        suppressBoundedSizeException={true}
        windowCorrectionConfig={
          {
            "applyToInitialOffset": true,
            "applyToItemScroll": true,
            "value": {
              "endCorrection": 0,
              "startCorrection": 0,
              "windowShift": -0,
            },
          }
        }
      >
        <View>
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <View
              style={
                {
                  "paddingLeft": 0,
                  "paddingTop": 0,
                }
              }
            />
            <View
              style={
                [
                  undefined,
                  undefined,
                ]
              }
            />
            <AutoLayoutView
              enableInstrumentation={false}
              horizontal={false}
              onBlankAreaEvent={[Function]}
              onLayout={[Function]}
              renderAheadOffset={0}
              scrollOffset={0}
              style={
                {
                  "height": 0,
                  "width": 0,
                }
              }
              windowSize={0}
            />
            <CellContainer
              index={-1}
              style={
                [
                  undefined,
                  undefined,
                ]
              }
            />
            <View
              style={
                {
                  "paddingBottom": 0,
                  "paddingRight": 0,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flex": 0.1,
        "justifyContent": "flex-end",
        "marginBottom": 80,
        "marginRight": 20,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Image
        source={1}
      />
    </View>
  </View>
</View>
`;
